### PR TITLE
Use resource.Quantity.AsApproximateFloat64() instead of AsInt64()

### PR DESCRIPTION
**Pull request by mcbenjemaa**
Towards https://github.com/giantswarm/giantswarm/issues/25909

Explantation: https://github.com/giantswarm/giantswarm/issues/25909#issuecomment-1433320441

### CODE REVIEW

**Comment by primeroz** [comment]
this did work on `glippy` and addressed both issues we had - https://github.com/giantswarm/roadmap/issues/2021#top

* VPA for prometheus is now created
```
➜ k get vpa -n glippy-prometheus
NAME         MODE   CPU   MEM        PROVIDED   AGE
prometheus   Auto   11m   52428800   True       2m2s
```
* HB is created ( so for some reason the failing VPA was breaking the reconcile loop entirely ) 
Screenshot from 2023-02-16 17-07-03


kudos @mcbenjemaa for the super quick turn-around on this

File: CHANGELOG.md [hunk 1]
```diff
@@ -7,6 +7,10 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ## [Unreleased]
 
+### Changed
+
+- Use resource.Quantity.AsApproximateFloat64() instead of AsInt64(), in order to avoid conversion issue when multiply cpu, e.g. 3880m
+
```
Comment (hervenicol) [inline CHANGELOG.md:13]: I'm not sure I understand the problematic case.
* It fails at multiplying `3880*0.9`? That's a round number - 3492
* It fails at multiplying `3.880x0.9`? Then, why it didn't fail when multiplying `4x0.9`?

File: service/controller/resource/monitoring/verticalpodautoscaler/resource.go [hunk 2]
```diff
@@ -212,13 +212,9 @@ func (r *Resource) getMaxMemory(ctx context.Context) (*resource.Quantity, error)
...
-	i, ok := q.AsInt64()
-	if !ok {
-		return nil, microerror.Maskf(quantityConversionError, q.String())
-	}
-
-	n := float64(i) * multiplier
 
+	i := q.AsApproximateFloat64()
+	n := i * multiplier
 	q.Set(int64(n))
```
Comment (hervenicol) [inline service/controller/resource/monitoring/verticalpodautoscaler/resource.go:218]: No need to `round()` to get an int from a float? (sorry, golang noob).

**Review by hervenicol — COMMENTED** [review]
I totally agree with the idea that we should use floats when multiplying by 0.x numbers.
I just don't understand how the example value fails. Personal curiosity.
And go noob questions.

**Review by hervenicol — APPROVED** [review]
LGTM if tested.

**Comment by mcbenjemaa** [comment]
**Tested:** 

- [x] glippy
- [x] gauss
- [x] gohan
- [x] golem

File: service/controller/resource/monitoring/verticalpodautoscaler/resource.go [hunk 3]
```diff
@@ -212,14 +212,10 @@ func (r *Resource) getMaxMemory(ctx context.Context) (*resource.Quantity, error)
...
-	if !ok {
-		return nil, microerror.Maskf(quantityConversionError, q.String())
-	}
-
-	n := float64(i) * multiplier
 
-	q.Set(int64(n))
+	i := q.AsApproximateFloat64()
+	n := i * multiplier
+	q.SetMilli(int64(n * 1000))
```
Comment (QuentinBisson) [inline service/controller/resource/monitoring/verticalpodautoscaler/resource.go:218]: Why use `SetMilli` to then multiply it by 1000 ?

Comment (hervenicol) [reply service/controller/resource/monitoring/verticalpodautoscaler/resource.go:218]: We have a number like `2.5`
`Set` requires cast to int => becomes `2`
and `SetMilli` uses milli-cpus, so expects `2500`.

Comment (QuentinBisson) [reply service/controller/resource/monitoring/verticalpodautoscaler/resource.go:218]: :ack:

**Pull request merged**